File: bridge/src/harbor_bridge/installer/runtime.py
"""
Runtime detection and management.

Detects available runtimes (Node.js, Python, Docker) and provides
information about how to install missing ones.
"""

import asyncio
import logging
import shutil
import subprocess
import sys
from dataclasses import dataclass
from enum import Enum
from pathlib import Path
from typing import Optional

logger = logging.getLogger(__name__)


class RuntimeType(str, Enum):
    """Types of runtimes we support."""
    NODE = "node"
    PYTHON = "python"
    DOCKER = "docker"


@dataclass
class Runtime:
    """Information about an installed runtime."""
    type: RuntimeType
    available: bool
    version: Optional[str] = None
    path: Optional[str] = None
    runner_cmd: Optional[str] = None  # e.g., "npx", "uvx", "docker"
    install_hint: Optional[str] = None
    
    def to_dict(self) -> dict:
        return {
            "type": self.type.value,
            "available": self.available,
            "version": self.version,
            "path": self.path,
            "runnerCmd": self.runner_cmd,
            "installHint": self.install_hint,
        }


class RuntimeManager:
    """Detects and manages runtimes for running MCP servers."""
    
    # Install hints for missing runtimes
    INSTALL_HINTS = {
        RuntimeType.NODE: {
            "darwin": "Install with: brew install node\nOr: https://nodejs.org/",
            "linux": "Install with: curl -fsSL https://deb.nodesource.com/setup_lts.x | sudo -E bash - && sudo apt-get install -y nodejs",
            "win32": "Download from: https://nodejs.org/",
        },
        RuntimeType.PYTHON: {
            "darwin": "Install with: brew install python\nOr use uv: curl -LsSf https://astral.sh/uv/install.sh | sh",
            "linux": "Install with: sudo apt install python3 python3-pip\nOr use uv: curl -LsSf https://astral.sh/uv/install.sh | sh",
            "win32": "Download from: https://python.org/\nOr use uv: powershell -c \"irm https://astral.sh/uv/install.ps1 | iex\"",
        },
        RuntimeType.DOCKER: {
            "darwin": "Install Docker Desktop: https://docker.com/products/docker-desktop/",
            "linux": "Install with: curl -fsSL https://get.docker.com | sh",
            "win32": "Install Docker Desktop: https://docker.com/products/docker-desktop/",
        },
    }
    
    def __init__(self):
        self._cache: dict[RuntimeType, Runtime] = {}
    
    async def detect_all(self, force_refresh: bool = False) -> list[Runtime]:
        """Detect all available runtimes."""
        if not force_refresh and len(self._cache) == 3:
            return list(self._cache.values())
        
        runtimes = await asyncio.gather(
            self.detect_node(),
            self.detect_python(),
            self.detect_docker(),
        )
        
        for runtime in runtimes:
            self._cache[runtime.type] = runtime
        
        return runtimes
    
    async def detect_node(self) -> Runtime:
        """Detect Node.js and npx availability."""
        runtime = Runtime(
            type=RuntimeType.NODE,
            available=False,
            install_hint=self._get_install_hint(RuntimeType.NODE),
        )
        
        # Check for node
        node_path = shutil.which("node")
        if node_path:
            runtime.path = node_path
            version = await self._get_version("node", "--version")
            if version:
                runtime.version = version
                runtime.available = True
        
        # Check for npx (preferred runner)
        npx_path = shutil.which("npx")
        if npx_path:
            runtime.runner_cmd = "npx"
        elif runtime.available:
            # Fallback to node with require
            runtime.runner_cmd = "node"
        
        self._cache[RuntimeType.NODE] = runtime
        return runtime
    
    async def detect_python(self) -> Runtime:
        """Detect Python and uvx/pip availability."""
        runtime = Runtime(
            type=RuntimeType.PYTHON,
            available=False,
            install_hint=self._get_install_hint(RuntimeType.PYTHON),
        )
        
        # Check for python
        python_path = shutil.which("python3") or shutil.which("python")
        if python_path:
            runtime.path = python_path
            version = await self._get_version(python_path, "--version")
            if version:
                runtime.version = version
                runtime.available = True
        
        # Check for uvx (preferred) or pipx
        uvx_path = shutil.which("uvx")
        if uvx_path:
            runtime.runner_cmd = "uvx"
        else:
            pipx_path = shutil.which("pipx")
            if pipx_path:
                runtime.runner_cmd = "pipx run"
            elif runtime.available:
                # Fallback to python -m
                runtime.runner_cmd = f"{python_path} -m"
        
        self._cache[RuntimeType.PYTHON] = runtime
        return runtime
    
    async def detect_docker(self) -> Runtime:
        """Detect Docker availability."""
        runtime = Runtime(
            type=RuntimeType.DOCKER,
            available=False,
            install_hint=self._get_install_hint(RuntimeType.DOCKER),
        )
        
        docker_path = shutil.which("docker")
        if docker_path:
            runtime.path = docker_path
            runtime.runner_cmd = "docker"
            
            # Check if Docker daemon is running
            try:
                result = subprocess.run(
                    ["docker", "info"],
                    capture_output=True,
                    timeout=5,
                )
                if result.returncode == 0:
                    runtime.available = True
                    version = await self._get_version("docker", "--version")
                    runtime.version = version
                else:
                    runtime.install_hint = "Docker is installed but not running. Start Docker Desktop."
            except (subprocess.TimeoutExpired, Exception):
                runtime.install_hint = "Docker is installed but not responding. Start Docker Desktop."
        
        self._cache[RuntimeType.DOCKER] = runtime
        return runtime
    
    def get_runtime(self, runtime_type: RuntimeType) -> Optional[Runtime]:
        """Get a cached runtime, or None if not detected yet."""
        return self._cache.get(runtime_type)
    
    def get_runtime_for_registry_type(self, registry_type: str) -> Optional[Runtime]:
        """Get the runtime needed for a registry type (npm, pypi, oci)."""
        mapping = {
            "npm": RuntimeType.NODE,
            "pypi": RuntimeType.PYTHON,
            "oci": RuntimeType.DOCKER,
        }
        runtime_type = mapping.get(registry_type.lower())
        if runtime_type:
            return self._cache.get(runtime_type)
        return None
    
    async def _get_version(self, cmd: str, *args: str) -> Optional[str]:
        """Get version string from a command."""
        try:
            result = subprocess.run(
                [cmd, *args],
                capture_output=True,
                text=True,
                timeout=5,
            )
            if result.returncode == 0:
                # Extract version from output (usually first line)
                output = result.stdout.strip() or result.stderr.strip()
                # Clean up common prefixes
                for prefix in ("v", "Python ", "Docker version "):
                    if output.startswith(prefix):
                        output = output[len(prefix):]
                return output.split()[0] if output else None
        except Exception as e:
            logger.debug(f"Failed to get version for {cmd}: {e}")
        return None
    
    def _get_install_hint(self, runtime_type: RuntimeType) -> str:
        """Get install hint for current platform."""
        platform = sys.platform
        if platform.startswith("linux"):
            platform = "linux"
        hints = self.INSTALL_HINTS.get(runtime_type, {})
        return hints.get(platform, hints.get("linux", ""))


# Singleton
_manager: Optional[RuntimeManager] = None


def get_runtime_manager() -> RuntimeManager:
    """Get the singleton RuntimeManager."""
    global _manager
    if _manager is None:
        _manager = RuntimeManager()
    return _manager

File: bridge-ts/src/installer/runtime.ts
/**
 * Runtime detection and management.
 * 
 * Detects available runtimes (Node.js, Python, Docker, Ollama).
 * Uses any-llm-ts for Ollama server availability checks.
 */

import { execSync } from 'node:child_process';
import { existsSync } from 'node:fs';
import { platform } from 'node:os';
import { log } from '../native-messaging.js';
import { Runtime, RuntimeType } from '../types.js';
import { AnyLLM } from '../any-llm-ts/src/index.js';

// Install hints for missing runtimes
const INSTALL_HINTS: Record<RuntimeType, Record<string, string>> = {
  [RuntimeType.NODE]: {
    darwin: 'Install with: brew install node\nOr: https://nodejs.org/',
    linux: 'Install with: curl -fsSL https://deb.nodesource.com/setup_lts.x | sudo -E bash - && sudo apt-get install -y nodejs',
    win32: 'Download from: https://nodejs.org/',
  },
  [RuntimeType.PYTHON]: {
    darwin: 'Install with: brew install python\nOr use uv: curl -LsSf https://astral.sh/uv/install.sh | sh',
    linux: 'Install with: sudo apt install python3 python3-pip\nOr use uv: curl -LsSf https://astral.sh/uv/install.sh | sh',
    win32: 'Download from: https://python.org/\nOr use uv: powershell -c "irm https://astral.sh/uv/install.ps1 | iex"',
  },
  [RuntimeType.DOCKER]: {
    darwin: 'Install Docker Desktop: https://docker.com/products/docker-desktop/',
    linux: 'Install with: curl -fsSL https://get.docker.com | sh',
    win32: 'Install Docker Desktop: https://docker.com/products/docker-desktop/',
  },
  [RuntimeType.OLLAMA]: {
    darwin: 'Install with: brew install ollama\nOr download from: https://ollama.com/',
    linux: 'Install with: curl -fsSL https://ollama.com/install.sh | sh',
    win32: 'Download from: https://ollama.com/',
  },
};

function which(cmd: string): string | null {
  try {
    const result = execSync(
      platform() === 'win32' ? `where ${cmd}` : `which ${cmd}`,
      { encoding: 'utf-8', timeout: 5000, stdio: ['pipe', 'pipe', 'pipe'] }
    );
    return result.trim().split('\n')[0] || null;
  } catch {
    return null;
  }
}

function getVersion(cmd: string, args: string[]): string | null {
  try {
    const result = execSync(`${cmd} ${args.join(' ')}`, { 
      encoding: 'utf-8', 
      timeout: 5000,
      stdio: ['pipe', 'pipe', 'pipe']
    });
    let version = result.trim();
    // Clean up common prefixes
    for (const prefix of ['v', 'Python ', 'Docker version ']) {
      if (version.startsWith(prefix)) {
        version = version.substring(prefix.length);
      }
    }
    return version.split(/\s+/)[0] || null;
  } catch {
    return null;
  }
}

function getInstallHint(type: RuntimeType): string {
  const p = platform();
  const hints = INSTALL_HINTS[type];
  return hints[p] || hints.linux || '';
}

export class RuntimeManager {
  private cache: Map<RuntimeType, Runtime> = new Map();

  async detectAll(forceRefresh: boolean = false): Promise<Runtime[]> {
    if (!forceRefresh && this.cache.size === 4) {
      return Array.from(this.cache.values());
    }

    const [node, python, docker, ollama] = await Promise.all([
      this.detectNode(),
      this.detectPython(),
      this.detectDocker(),
      this.detectOllama(),
    ]);

    return [node, python, docker, ollama];
  }

  async detectNode(): Promise<Runtime> {
    const runtime: Runtime = {
      type: RuntimeType.NODE,
      available: false,
      version: null,
      path: null,
      runnerCmd: null,
      installHint: getInstallHint(RuntimeType.NODE),
    };

    const nodePath = which('node');
    if (nodePath) {
      runtime.path = nodePath;
      runtime.version = getVersion('node', ['--version']);
      if (runtime.version) {
        runtime.available = true;
      }
    }

    const npxPath = which('npx');
    if (npxPath) {
      runtime.runnerCmd = 'npx';
    } else if (runtime.available) {
      runtime.runnerCmd = 'node';
    }

    this.cache.set(RuntimeType.NODE, runtime);
    return runtime;
  }

  async detectPython(): Promise<Runtime> {
    const runtime: Runtime = {
      type: RuntimeType.PYTHON,
      available: false,
      version: null,
      path: null,
      runnerCmd: null,
      installHint: getInstallHint(RuntimeType.PYTHON),
    };

    const pythonPath = which('python3') || which('python');
    if (pythonPath) {
      runtime.path = pythonPath;
      runtime.version = getVersion(pythonPath, ['--version']);
      if (runtime.version) {
        runtime.available = true;
      }
    }

    // Check for uvx (preferred) or pipx
    const uvxPath = which('uvx');
    if (uvxPath) {
      runtime.runnerCmd = 'uvx';
    } else {
      const pipxPath = which('pipx');
      if (pipxPath) {
        runtime.runnerCmd = 'pipx run';
      } else if (runtime.available) {
        runtime.runnerCmd = `${pythonPath} -m`;
      }
    }

    this.cache.set(RuntimeType.PYTHON, runtime);
    return runtime;
  }

  async detectDocker(): Promise<Runtime> {
    const runtime: Runtime = {
      type: RuntimeType.DOCKER,
      available: false,
      version: null,
      path: null,
      runnerCmd: null,
      installHint: getInstallHint(RuntimeType.DOCKER),
    };

    const dockerPath = which('docker');
    if (dockerPath) {
      runtime.path = dockerPath;
      runtime.runnerCmd = 'docker';

      // Check if Docker daemon is running
      try {
        execSync('docker info', { 
          timeout: 5000,
          stdio: ['pipe', 'pipe', 'pipe']
        });
        runtime.available = true;
        runtime.version = getVersion('docker', ['--version']);
      } catch {
        runtime.installHint = 'Docker is installed but not running. Start Docker Desktop.';
      }
    }

    this.cache.set(RuntimeType.DOCKER, runtime);
    return runtime;
  }

  async detectOllama(): Promise<Runtime> {
    const runtime: Runtime = {
      type: RuntimeType.OLLAMA,
      available: false,
      version: null,
      path: null,
      runnerCmd: null,
      installHint: getInstallHint(RuntimeType.OLLAMA),
    };

    const ollamaPath = which('ollama');
    if (ollamaPath) {
      runtime.path = ollamaPath;
      runtime.runnerCmd = 'ollama';
      runtime.version = getVersion('ollama', ['--version']);

      // Use any-llm-ts to check if Ollama server is running
      try {
        const ollama = AnyLLM.create('ollama');
        const isRunning = await ollama.isAvailable();
        
        if (isRunning) {
          runtime.available = true;
        } else {
          // Ollama installed but not running - still mark as available since it can be started
          runtime.available = true;
          runtime.installHint = 'Ollama is installed. Run "ollama serve" to start the server.';
        }
      } catch {
        // Ollama installed but server check failed - still mark as available
        runtime.available = true;
        runtime.installHint = 'Ollama is installed. Run "ollama serve" to start the server.';
      }
    }

    this.cache.set(RuntimeType.OLLAMA, runtime);
    log(`[RuntimeManager] Ollama: ${runtime.available ? 'available' : 'not found'} (${runtime.version || 'unknown version'})`);
    return runtime;
  }

  getRuntime(type: RuntimeType): Runtime | undefined {
    return this.cache.get(type);
  }

  getRuntimeForRegistryType(registryType: string): Runtime | undefined {
    const mapping: Record<string, RuntimeType> = {
      npm: RuntimeType.NODE,
      pypi: RuntimeType.PYTHON,
      oci: RuntimeType.DOCKER,
    };
    const type = mapping[registryType.toLowerCase()];
    return type ? this.cache.get(type) : undefined;
  }
}

// Singleton
let _manager: RuntimeManager | null = null;

export function getRuntimeManager(): RuntimeManager {
  if (!_manager) {
    _manager = new RuntimeManager();
  }
  return _manager;
}






